Mathlib/AlgebraicTopology/SimplexCategory/GeneratorsRelations/EpiMono.lean

```diff
@@ -98,4 +96,182 @@ lemma eq_or_len_le_of_P_δ {x y : SimplexCategoryGenRel} {f : x ⟶ y} (h_δ : P
+
+/-- A low-dimensional special case of the previous -/
+private lemma switch_δ_σ₀ (i : Fin 1) (i' : Fin 2) :
+    δ i' ≫ σ i = 𝟙 _ := by
+  rcases i with ⟨i, hi⟩
+  rcases i' with ⟨i', hi'⟩
+  simp at hi hi'
+  rw [Nat.lt_iff_le_pred Nat.zero_lt_two] at hi'
+  simp at hi'
+  subst hi
+  obtain h | h := Nat.le_one_iff_eq_zero_or_eq_one.mp hi'
+  · subst h
+    simp only [Fin.zero_eta, Fin.isValue, ← Fin.castSucc_zero, δ_comp_σ_self]
+  · subst h
+    simp only [Fin.mk_one, Fin.isValue, Fin.zero_eta]
+    rw [← Fin.succ_zero_eq_one, δ_comp_σ_succ]
+
+private lemma factor_δ_σ {n : ℕ} (i : Fin (n + 1)) (i' : Fin (n + 2)) :
+    ∃ (z : SimplexCategoryGenRel) (e : mk n ⟶ z) (m : z ⟶ mk n)
+      (_ : P_σ e) (_ : P_δ m), δ i' ≫ σ i = e ≫ m := by
+  cases n with
+  | zero =>
+    rw [switch_δ_σ₀]
+    use mk 0, 𝟙 _, 𝟙 _, P_σ.id_mem _, P_δ.id_mem _
+    simp
+  | succ n =>
+    obtain h | ⟨j, j', h⟩ := switch_δ_σ i i' <;> rw [h]
+    · use mk (n + 1), 𝟙 _, 𝟙 _, P_σ.id_mem _, P_δ.id_mem _
+      simp
+    · use mk n, σ j, δ j', P_σ.σ _, P_δ.δ _
+
+/-- An auxiliary lemma that shows there exists a factorization as a P_δ followed by a P_σ for
+morphisms of the form `P_δ ≫ σ`. -/
+private lemma factor_P_δ_σ {n : ℕ} (i : Fin (n + 1)) {x : SimplexCategoryGenRel}
+    (f : x ⟶ mk (n + 1)) (hf : P_δ f) : ∃ (z : SimplexCategoryGenRel) (e : x ⟶ z) (m : z ⟶ mk n)
+      (_ : P_σ e) (_ : P_δ m), f ≫ σ i = e ≫ m := by
+  induction n using Nat.case_strong_induction_on generalizing x with
+  | hz => cases hf with
+    | of _ h => cases h; exact factor_δ_σ _ _
+    | id  =>
+      rw [Category.id_comp]
+      use mk 0, σ i, 𝟙 _, P_σ.σ _, P_δ.id_mem _
+      simp
+    | comp_of j f hf hg =>
+      cases hg
+      obtain ⟨h', hf'⟩ | hf' := eq_or_len_le_of_P_δ hf
+      · subst h'
+        simp only [eqToHom_refl] at hf'
+        subst hf'
+        rw [Category.id_comp]
+        exact factor_δ_σ _ _
+      · simp at hf'
+  | hi n h_rec =>
+    cases hf with
+    | of _ h => cases h; exact factor_δ_σ _ _
+    | @id n =>
+      rw [Category.id_comp]
+      use mk (n + 1), σ i, 𝟙 _, P_σ.σ _, P_δ.id_mem _
+      simp
+    | comp_of f g hf hg =>
+      obtain ⟨h', h''⟩ | h := eq_or_len_le_of_P_δ hf
+      · subst h'
+        cases hg
+        rw [eqToHom_refl] at h''; subst h''
+        rw [Category.id_comp]
+        exact factor_δ_σ _ _
+      · have hg' := hg
+        rcases hg' with ⟨i'⟩
+        obtain h' | ⟨j, j', h'⟩ := switch_δ_σ i i' <;> rw [Category.assoc, h']
+        · rw [Category.comp_id]
+          use x, 𝟙 x, f, P_σ.id_mem _, hf
+          simp
+        · rw [mk_len, Nat.lt_add_one_iff] at h
+          obtain ⟨z, e, m₁, he, hm₁, h⟩ := h_rec n (Nat.le_refl _) j f hf
+          rw [reassoc_of% h]
+          use z, e, m₁ ≫ δ j', he, P_δ.comp_mem _ _ hm₁ (P_δ.δ _)
+
+/-- Any morphism in `SimplexCategoryGenRel` can be decomposed as a `P_σ` followed by a `P_δ`. -/
+theorem exists_P_σ_P_δ_factorization {x y : SimplexCategoryGenRel} (f : x ⟶ y) :
```

Comment:
It could be nice to state this also as `HasFactorization P_σ P_δ` (see `MorphismProperty.Factorization`).